src/libsyntax/ext/expand.rs

```diff
@@ -755,6 +755,15 @@ fn expand_annotatable(a: Annotatable,
                 }
                 result.into_iter().map(|i| Annotatable::Item(i)).collect()
             },
+            ast::ItemKind::ExternCrate(_) => {
+                // We need to error on `#[macro_use] extern crate` when it isn't at the
+                // crate root, because `$crate` won't work properly.
+                let allows_macros = fld.cx.syntax_env.is_crate_root();
+                for def in fld.cx.loader.load_crate(&it, allows_macros) {
+                    fld.cx.insert_macro(def);
+                }
+                SmallVector::one(Annotatable::Item(it))
```

Comment:
Before this PR, `new_attrs` wasn't attached back (so the expanded decorator attributes weren't removed) when folding modules, trait items, and impl items, so I don't think it matters much.

That being said, I think the decorator attributes should always be removed since macro expansion should be idempotent, so I added a commit that always reattaches `new_attrs`.